### DISCUSSION TITLE
:has Selector or equivariant?

### DISCUSSION THREAD

**Discussion by Omsad** [AvaloniaUI/Avalonia#17310] [Q&A] [answered]
I have an ItemsControl which contains header and content data templates, e.g. 

````axaml
        ...
        <ItemsControl.ItemTemplate>
          <MultipleDataTemplateSelector>
            <DataTemplate DataType="model:HeaderItemViewModel">
              <Grid Classes="header-item">
                ...
              </Grid>
            </DataTemplate>
            <DataTemplate DataType="model:ContentItemViewModel">
              <Grid Classes="content-item">
                ...
              </Grid>
            </DataTemplate>
          </MultipleDataTemplateSelector>
        </ItemsControl.ItemTemplate>
        ...
````

I would like to alternate the background colour of the content items. I currently have the following styles:
````axaml
<Style Selector="ContentPresenter Grid">
  <Setter Property="Background" Value="{StaticResource Transparent}" />
</Style>
<Style Selector="ContentPresenter:nth-child(2n) Grid.content-item">
  <Setter Property="Background" Value="{StaticResource SecondaryTertiary}" />
</Style>
````

This works fine apart from where there is a content-item => header-item => content-Item. This causes the content-item's to have the same background colour, e.g. the header-item is included in the :nth-child(2n) so the index of is even.

Is there a way of ignoring any header item when calculating the nth child count, as in CSS I think I could do :
````axaml
<Style Selector="ContentPresenter:not(:has(> Grid.header-item)):nth-child(2n) Grid.content-item">
  <Setter Property="Background" Value="{StaticResource SecondaryTertiary}" />
</Style>
````

Is there an equivariant **:has** selector?

**Comment by maxkatz6** [comment]
There is no such selector on Avalonia. But it would be a useful addition.
If you want, can you create a feature request issue?
We unlikely to prioritize it soon, but it would be a good start to have a tracking issue.

**Comment by maxkatz6** [comment]
> ContentPresenter:not(:has(> Grid.header-item)):nth-child(2n)

Btw, this selector wouldn't work either way (at least, as intended), even if Avalonia supported ":has". Just like it won't work in CSS.
Because ":nth-child" ignores filtration by other selectors for the element.

**Reply by maxkatz6** [reply, marked as answer]
There is actually a new selector syntax in CSS, which would solve your task even without `:has` - https://www.youtube.com/watch?v=AerrBBdLlDc

**Reply by maxkatz6** [reply]
Other than this, I don't think your task can be solved with Avalonia styles alone right now. You would need to mix it with C# logic.

**Reply by Omsad** [reply]
Okay, thank you. 

I will likely leave this for the moment then and come back and have a look at extending the nth-child selector at a later point, as it was a nice to have rather than must.
